lib/sshutils/x11/auth.go

```diff
@@ -0,0 +1,221 @@
+	x.Cmd.Args = append(x.Cmd.Args, "remove", display)
+	return trace.Wrap(x.run())
+}
+
+// AddEntry runs "xauth add" to add the given xauth entry.
+func (x *XAuthCommand) AddEntry(entry *XAuthEntry) error {
+	x.Cmd.Args = append(x.Cmd.Args, "add", entry.Display, entry.Proto, entry.Cookie)
+	return trace.Wrap(x.run())
+}
+
+// GenerateUntrustedCookie runs "xauth generate untrusted" to create a new untrusted xauth
+// entry for the given display. A timeout can optionally be set for the xauth entry.
+//
+// An untrusted cookie will signal to the XServer that fewer X privileges should be provided
+// when opening local connections with this cookie. This prevents attackers from using the
+// cookie to perform actions like keystroke monitoring.
+func (x *XAuthCommand) GenerateUntrustedCookie(display, proto string, timeout uint) error {
+	x.Cmd.Args = append(x.Cmd.Args, "generate", "untrusted", display, proto)
+	if timeout != 0 {
+		// Add some slack to the ttl to avoid XServer from denying
+		// access to the ssh session during its lifetime.
+		var timeoutSlack uint = 60
+		x.Cmd.Args = append(x.Cmd.Args, "timeout", fmt.Sprint(timeout+timeoutSlack))
+	}
+	return trace.Wrap(x.run())
+}
+
+// run Run and wrap error with stderr.
+func (x *XAuthCommand) run() error {
+	err := x.Cmd.Run()
+	if err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			return trace.Wrap(err, "stderr: %q", exitErr.Stderr)
+		}
+	}
+	return trace.Wrap(err)
+}
+
+// run Output and wrap error with stderr.
+func (x *XAuthCommand) output() ([]byte, error) {
+	out, err := x.Cmd.Output()
+	if err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			return nil, trace.Wrap(err, "stderr: %q", exitErr.Stderr)
+		}
+	}
+	return out, trace.Wrap(err)
+}
+
+// ReadAndRewriteXAuthPacket reads the initial xauth packet from the xserver request. The xauth packet has 2 parts:
+//  1. fixed size buffer (12 bytes) - holds byteOrder bit, and the sizes of the protocol string and auth data
+//  2. variable size xauth packet - holds xauth protocol and data used to connect to the remote XServer.
+//
+// Then it compares the received auth packet with the auth proto and fake cookie
+// sent to the server with the original "x11-req". If the data matches, the auth
+// packet is returned with the fake cookie replaced by the real cookie to provide
+// access to the client's X display.
+func ReadAndRewriteXAuthPacket(xchan ssh.Channel, authProto, fakeCookie, realCookie string) ([]byte, error) {
+	// xauth packet starts with a fixed sized buffer of 12 bytes
+	// which is used to size and decode the remaining bytes
+	initBufSize := 12
+	initBuf := make([]byte, initBufSize)
+	if _, err := io.ReadFull(xchan, initBuf); err != nil {
+		return nil, trace.Wrap(err, "x11 channel initial packet buffer missing or too short")
+	}
+
+	var protoLen, dataLen int
+	switch byteOrder := initBuf[0]; byteOrder {
+	///* Byte order MSB first. */
+	case 0x42:
```

Comment:
make constants for these byte orders?

Reply:
Changed these to 'B' and 'l' and simplified the switch statement.